pkg/duckdb/ephemeral_db.go

```diff
@@ -5,58 +5,659 @@ package duck
+			}
+		case sql.Scanner:
+			// Support sql.Scanner interface for custom types
+			if err := d.Scan(value); err != nil {
+				return fmt.Errorf("error scanning column %d: %w", i, err)
+			}
+		default:
+			return fmt.Errorf("unsupported destination type for column %d: %T", i, dest[i])
+		}
+	}
+
+	return nil
+}
+
+func (r *adbcRows) Columns() ([]string, error) {
+	schema := r.reader.Schema()
+	columns := make([]string, schema.NumFields())
+	for i := range schema.NumFields() {
+		columns[i] = schema.Field(i).Name
 	}
-	defer func(conn *sqlx.DB) {
-		if err := conn.Close(); err != nil {
-			panic(err)
+	return columns, nil
+}
+
+func (r *adbcRows) ColumnTypes() ([]ColumnType, error) {
+	schema := r.reader.Schema()
+	types := make([]ColumnType, schema.NumFields())
+	for i := range schema.NumFields() {
+		field := schema.Field(i)
+		// Map Arrow type names to SQL type names for compatibility
+		sqlType := arrowTypeToSQLType(field.Type.String())
+		types[i] = &adbcColumnType{name: field.Name, dbType: sqlType}
+	}
+	return types, nil
+}
+
+// arrowTypeToSQLType maps Arrow type names to SQL type names.
+func arrowTypeToSQLType(arrowType string) string {
+	switch arrowType {
+	case "int8", "int16", "int32", "int64":
+		return "INTEGER"
+	case "uint8", "uint16", "uint32", "uint64":
+		return "INTEGER"
+	case "float32", "float64":
+		return "DOUBLE"
+	case "utf8", "large_utf8":
+		return "VARCHAR"
+	case "binary", "large_binary":
+		return "BLOB"
+	case "bool":
+		return "BOOLEAN"
+	case "date32", "date64":
+		return "DATE"
+	case "timestamp":
+		return "TIMESTAMP"
+	default:
+		// For decimal types, uppercase and fix spacing: "decimal(5, 2)" -> "DECIMAL(5,2)"
+		if len(arrowType) >= 7 && arrowType[:7] == "decimal" {
+			// Replace "decimal" with "DECIMAL" and remove spaces after commas
+			result := "DECIMAL" + arrowType[7:]
+			// Remove spaces after commas
+			result = strings.ReplaceAll(result, ", ", ",")
+			return result
+		}
+		// For timestamp with unit, just return "TIMESTAMP": "timestamp[us]" -> "TIMESTAMP"
+		if len(arrowType) >= 9 && arrowType[:9] == "timestamp" {
+			return "TIMESTAMP"
+		}
+		// For other types, keep the original
+		return arrowType
+	}
+}
+
+func (r *adbcRows) Err() error {
+	return r.err
+}
+
+func (r *adbcRows) Close() error {
+	if r.closed {
+		return nil
+	}
+	r.closed = true
+
+	// Release all retained records
+	for _, rec := range r.allRecords {
+		rec.Release()
+	}
+	r.allRecords = nil
+
+	r.reader.Release()
+	r.stmt.Close()
+	_ = r.conn.Close()
+	_ = r.db.Close()
+	return nil
+}
+
+// adbcRow implements the Row interface.
+type adbcRow struct {
+	values []interface{}
+	err    error
+}
+
+func (r *adbcRow) Scan(dest ...interface{}) error {
+	if r.err != nil {
+		return r.err
+	}
+
+	if len(dest) != len(r.values) {
+		return fmt.Errorf("expected %d destination arguments, got %d", len(r.values), len(dest))
+	}
+
+	for i, value := range r.values {
```

Comment:
Is this logic being duplicated? Can it be factored out to a helper function? (Or eliminated entirely by using the sqldriver package?)